File: 04 architecture/02 accumulator machine.rb
#!/usr/bin/env ruby

require_relative "vm"

class Adder < VM
	def interpret
		@s = []
		@a = 0
		super
	end

	def clear
		@a = 0
		read_program.call
	end

	def push_value
		@s.push(read_program)
		read_program.call
	end

	def push
		@s.push(@accum)
		read_program.call
	end

	def add
		@a += @s.pop
		read_program.call
	end

	def jump_if_not_zero
		if @a == 0
			@pc += 1
		else
			@pc = @program[@pc]
		end
		read_program.call
	end

	def exit
		throw :program_complete
	end

	def print_state
		puts "#{@pc}: @a = #{@a}, @s => #{@s}"
	end
end

Adder.new(
	:clear,
	:push_value, 13,
	:print_state,
	:add,
	:print_state,
	:push_value, -1,
	:print_state,
	:add,
	:print_state,
	:read_program,
	:print_state,
	:jump_if_not_zero, 6,
	:exit
).interpret

File: 04 architecture/02 accumulator machine.go
package main

import "fmt"
import "os"

type AccMachine struct {
	A int
	VM
}

func (a *AccMachine) String() string {
	return fmt.Sprintf("%v, A=%v", a.VM, a.A)
}

func (a *AccMachine) Clear() {
	a.A = 0
}

func (a *AccMachine) LoadValue() {
	a.PC++
	a.A = a.read_program().(int)
}

func (a *AccMachine) Add() {
	a.PC++
	a.A += a.read_program().(int)
}

func (a *AccMachine) JumpIfNotZero() {
	a.PC++
	if a.A != 0 {
		a.PC = a.m[a.PC].(int)
	}
}

func main() {
	a := new(AccMachine)
	print_state := func() {
		fmt.Println(a)
	}
	skip := func() { a.PC++ }
	a.Load(
		a.Clear,
		a.LoadValue, 13,
		Label("decrement"),
		a.Add, -1,
		print_state,
		skip,
		print_state,
		a.JumpIfNotZero, Label("decrement"),
		print_state,
		func() { os.Exit(0) },
	).Run()
}